### PR TITLE
docs: convert `ErrorHandlingRationale` from `.rst` to `.md`

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Resolves partially #49997.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

File: docs/ErrorHandlingRationale.md [hunk 1]
```diff
@@ -0,0 +1,1940 @@
...
+    half-a-dozen methods like this:
+
+```objc
+- (BOOL)sessionDriver:(ISyncSessionDriver *)sender
+    didRegisterClientAndReturnError:(NSError **)outError;
+```
+
+    Fortunately, these delegate methods were all deprecated in Lion, and
+    Swift currently doesn't even import deprecated methods.
+
```
Comment (AnthonyLatsis) [inline docs/ErrorHandlingRationale.md:1846]: Underindented code block.

File: docs/ErrorHandlingRationale.rst [hunk 2]
```diff
@@ -1,2041 +0,0 @@
-Error Handling Rationale and Proposal
-*************************************
-
-.. contents::
```
Comment (AnthonyLatsis) [inline docs/ErrorHandlingRationale.rst:4]: The markdown file needs a manually written index. In `.rst` files, it is rendered automatically using this directive.

**Review by AnthonyLatsis — APPROVED** [review]

File: docs/ErrorHandlingRationale.md (hunk 1 above)
Comment (MaxDesiatov) [reply docs/ErrorHandlingRationale.md:1846]: I had indented code blocks in one of the previous PRs, and I think you said that those don't render correctly, didn't you?

Comment (AnthonyLatsis) [reply docs/ErrorHandlingRationale.md:1846]: Yeah, but those were over-indented relative to the text. This is the opposite case — the code block belongs to the bullet. Check out how it renders.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test and merge

File: docs/ErrorHandlingRationale.md [hunk 3]
```diff
@@ -1,5 +1,62 @@
 # Error Handling Rationale and Proposal
 
+- [Error Handling Rationale and Proposal](#error-handling-rationale-and-proposal)
```
Comment (AnthonyLatsis) [inline docs/ErrorHandlingRationale.md:3]: I don’t think this one is necessary.

Comment (AnthonyLatsis) [reply docs/ErrorHandlingRationale.md:3]: While the tests are failing I will take this opportunity to suggest adding a heading to the table, as in [GettingStarted.md](https://github.com/apple/swift/blob/main/docs/HowToGuides/GettingStarted.md) or [ARCOptimization.md](https://github.com/apple/swift/blob/main/docs/ARCOptimization.md)

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test and merge

**Pull request merged**